include/swift/Frontend/FrontendOptions.h

```diff
@@ -68,33 +68,37 @@ enum class InputFileKind {
 
 /// Information about all the inputs to the frontend.
 class FrontendInputs {
+  friend class ArgsToFrontendInputsConverter;
+
 private:
   /// The names of input files to the frontend.
   std::vector<std::string> InputFilenames;
 
   /// Input buffers which may override the file contents of input files.
   std::vector<llvm::MemoryBuffer *> InputBuffers;
 
-  /// The input for which output should be generated. If not set, output will
-  /// be generated for the whole module.
-  Optional<SelectedInput> PrimaryInput;
+  /// The inputs for which output should be generated. If empty, output will
+  /// be generated for the whole module, in other words, whole-module mode.
+  /// Even if every input file is mentioned here, it is not the same as
+  /// whole-module mode.
+  std::vector<SelectedInput> PrimaryInputs;
 
 public:
   // Readers:
 
   // Input filename readers
   ArrayRef<std::string> getInputFilenames() const { return InputFilenames; }
-  bool hasInputFilenames() const { return !getInputFilenames().empty(); }
+  bool haveInputFilenames() const { return !getInputFilenames().empty(); }
```

Comment:
This is *not important* enough to block merging, but I thought about it and I think `has` is still correct. The context we're looking at is

```c++
// if (the) frontend input info has (any) input filenames…
if (frontendInputInfo.hasInputFilenames())
```

That is, `has` vs. `have` applies to the plurality of `this`. I see that the type has a plural name, but I don't think that means values of the type are really plural. It's not a collection (no `begin()` and `end()`), and even if it were I think the collection itself would still be non-plural.

Again, this is not important. You could still merge this. But maybe going forward…